napari/utils/actions/_types.py

```diff
@@ -0,0 +1,292 @@
+from __future__ import annotations
+
+import os
+import sys
+from functools import cached_property
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Generic,
+    List,
+    NamedTuple,
+    NewType,
+    Optional,
+    TypedDict,
+    TypeVar,
+    Union,
+)
+
+from pydantic import BaseModel, Field
+
+from ...utils import context
+from ...utils.translations import TranslationString
+from ._menus import MenuId
+
+WINDOWS = os.name == 'nt'
+MACOS = sys.platform == 'darwin'
+LINUX = sys.platform.startswith("linux")
+
+TranslationOrStr = Union[TranslationString, str]
+CommandId = NewType("CommandId", str)
+KeyCode = NewType("KeyCode", str)
+IconCode = NewType("IconCode", str)
+CommandHandler = TypeVar("CommandHandler", bound=Callable[..., Any])
+
+if TYPE_CHECKING:
+
+    # Typed dicts mimic the API of their pydantic counterparts.
+    # Since pydantic allows you to pass in either an object or a dict,
+    # This lets us use either anywhere, without losing typing support.
+    # e.g. Union[MenuRuleDict, MenuRule]
+    class MenuRuleDict(TypedDict, total=False):
+        when: Optional[context.Expr]
+        group: str
+        order: Optional[float]
+        id: MenuId
+
+    class KeybindingRuleDict(TypedDict, total=False):
+        primary: Optional[KeyCode]
+        win: Optional[KeyCode]
+        linux: Optional[KeyCode]
+        mac: Optional[KeyCode]
+        weight: int
+        when: Optional[context.Expr]
+
+
+# ------------------ commands-related types --------------------
+
+
+class Icon(BaseModel):
+    """Icons used to represent commands, or submenus.
+
+    May provide both a light and dark variant.  If only one is provided, it is used
+    in all theme types.
+    """
+
+    dark: Optional[IconCode] = Field(
+        None,
+        description="Icon path when a dark theme is used. These may be superqt "
+        "fonticon keys, such as `fa5s.arrow_down`",
+    )
+    light: Optional[IconCode] = Field(
+        None,
+        description="Icon path when a light theme is used. These may be superqt "
+        "fonticon keys, such as `fa5s.arrow_down`",
+    )
```

Comment:
I do not understand this. Napari allows defining custom themes. how we could distinguish if the theme is dark? and when using font awesome icons when do need to define different icons for different themes?

Reply:
> how we could distinguish if the theme is dark?

Because it's in the theme contribution api `contributions.themes.type`:  
https://napari.org/plugins/contributions.html#contributions-themes


> and when using font awesome icons when do need to define different icons for different themes?

don't "need" to.  If we wanted to, then that differentiation would go elsewhere, (in the theme).  For how vscode does it, see [contributes.icons](https://code.visualstudio.com/api/references/contribution-points#contributes.icons) and [Product Icon Reference](https://code.visualstudio.com/api/references/icons-in-labels)

But note also that there will be some levels of reduction as you go from "how the plugin declares it" to "how it's internally registered".  For the purpose here, we are mostly acknowledging that an icon _can_ have a light and dark version